tests/functional/syntax/modules/test_initializers.py

```diff
@@ -0,0 +1,942 @@
+"""
+tests for the uses/initializes checker
+main properties to test:
+- state usage -- if a module uses state, it must `used` or `initialized`
+- conversely, if a module does not touch state, it should not be `used`
+- global initializer check: each used module is `initialized` exactly once
+"""
+
+import pytest
+
+from vyper.compiler import compile_code
+from vyper.exceptions import (
+    BorrowException,
+    ImmutableViolation,
+    InitializerException,
+    StructureException,
+    UndeclaredDefinition,
+)
+
+
+def test_initialize_uses(make_input_bundle):
+    lib1 = """
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+    """
+    lib2 = """
+import lib1
+
+uses: lib1
+
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+
+@internal
+def foo():
+    lib1.counter += 1
+    """
+    main = """
+import lib2
+import lib1
+
+initializes: lib2[lib1 := lib1]
+initializes: lib1
+
+@deploy
+def __init__():
+    lib1.__init__()
+    lib2.__init__()
+    """
+    input_bundle = make_input_bundle({"lib1.vy": lib1, "lib2.vy": lib2})
+
+    assert compile_code(main, input_bundle=input_bundle) is not None
+
+
+def test_initialize_uses_attribute(make_input_bundle):
+    lib1 = """
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+    """
+    lib2 = """
+import lib1
+
+uses: lib1
+
+counter: uint256
+
+@deploy
+def __init__():
+    pass
+
+@internal
+def foo():
+    lib1.counter += 1
+    """
+    main = """
+import lib1
+import lib2
+
+initializes: lib2[lib1 := lib1]
+initializes: lib1
+
+@deploy
+def __init__():
+    lib2.__init__()
+    # demonstrate we can call lib1.__init__ through lib2.lib1
+    # (not sure this should be allowed, really.
```

Comment:
Yeah this seems like a bad practice to allow

Reply:
idk -- i think it's bad practice, but i'm not convinced it's worth promoting to a semantic error. seems more like a linter error.